elasticsearch/es-config.py

```diff
@@ -0,0 +1,252 @@
+
+
+def _vm_memory(ram_string):
+  # Try to keep this simple.
+  # We only support a string that ends with "gb".
+  #
+  # Tokenize the value and return a vm_memory value as per:
+  #
+  # https://cloud.google.com/compute/docs/instances/creating-instance-with-custom-machine-type
+  # NOTE: memory must be a multiple of 256 MB
+  #
+  # So just convert GB to MB.
+
+  if ram_string.endswith('GB'):
+    return int(float(ram_string[:-2]) * 1024)
+  else:
+    gen_util.err_exit(1, [
+      "Invalid 'ram' value: '{ram_string}'",
+      "Must end in 'GB'"])
+
+
+def _machine_type(node_config):
+  cpu = node_config.get('vm_cpu', 2)
+  mem = _vm_memory(node_config.get('vm_ram', '6gb'))
+
+  # Default to e2 for now; might want to make this configurable
+  return f"e2-custom-{cpu}-{mem}"
+
+
+def k8_create(config):
+  cluster_name = config['name']
+
+  # Ensure the deployments directory is clean for this cluster
+  gen_util.rm_deployment_files(cluster_name)
+
+  cluster_config = {
+    'name': config['name'],
+    'project': config['project'],
+    'zone': config['zone']
+  }
+
+  master_pool_config = {
+    'name': 'master-pool',
+    'count': config['master']['count'],
+    'machine-type': _machine_type(config['master']),
+    'node-type': 'es-master'
+  }
+
+  data_pool_config = {
+    'name': 'data-pool',
+    'count': config['data']['count'],
+    'machine-type': _machine_type(config['data']),
+    'node-type': 'es-data'
+  }
+
+  # The gcloud create cluster command doesn't allow one to name the default
+  # pool (which we'd like to use for the master), though there is a feature
+  # request filed for this.
+  # For now, we create a default cluster, add the master and data pools
+  # and then delete the default-pool.
+
+  gen_util.timestamp()
+  print(f"*** Create Kubernetes cluster {cluster_name}.*")
+  k8_util.create_cluster(cluster_config)
+
+  gen_util.timestamp()
+  k8_util.create_node_pool(cluster_config, master_pool_config)
+  gen_util.timestamp()
+  k8_util.create_node_pool(cluster_config, data_pool_config)
+
+  gen_util.timestamp()
+  k8_util.delete_node_pool(cluster_config, { 'name': 'default-pool' }, True)
+
+  gen_util.timestamp()
+
+
+def k8_delete(config):
+  cluster_name = config['name']
+
+  # Delete the cluster
+  gen_util.timestamp()
+  print(f"*** Delete Kubernetes cluster {cluster_name}.*")
+  k8_util.delete_cluster(config)
+
+  # Clean up the deployments directory
+  gen_util.timestamp()
+  print(f"*** Delete deployments files for {cluster_name}")
+  gen_util.rm_deployment_files(cluster_name)
+
+  gen_util.timestamp()
+
+
+def get_loadbalancer_ip(config):
+
+  print("Getting elasticsearch loadbalancer IP address")
+
+  # Right after deployment, the service will not yet be defined
+  # Allow for multiple attemps.
+
+  attempt_max = 12
+  for attempt in range(attempt_max):
+    # NOTE: The hard-coded service name "elasticsearch-es-http" likely
+    # limits deployments to one per cloud project. This limitation
```

Comment:
I'm not sure this is true. One cloud project can have multiple k8s clusters. You would need to have your kubectl context set to the correct cluster, each of which has its own service named elasticsearch-es-http.

We had this [setup-gcloud.sh](https://github.com/DataBiosphere/data-explorer-indexers/blob/master/util/setup-gcloud.sh) script that sets the kubectl context to point to the most recently created cluster.

Reply:
I think you are right. Removed the comment.